lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java

```diff
@@ -2876,4 +2875,139 @@ public void onError(ErrorInfo reason) {
 		}
 	}
 
+	/**
+	 * Verify protocol messages sent on Presence.enter() follow specs if sent from correct state and
+	 * the call fails if sent from DETACHED state
+	 *
+	 * Tests RTP8c, RTP8g
+	 */
+	@Test
+	public void protocol_enter_message_format() throws AblyException, InterruptedException {
+		AblyRealtime ably = null;
+		String oldTransport = Defaults.TRANSPORT;
+
+		try {
+			final ArrayList<PresenceMessage> sentPresence = new ArrayList<>();
+			Defaults.TRANSPORT = MockWebsocketFactory.class.getName();
+			/* Allow send but record all the presence messages for later analysis */
+			MockWebsocketFactory.allowSend(new MockWebsocketFactory.MessageFilter() {
+				@Override
+				public boolean matches(ProtocolMessage message) {
+					if (message.action == ProtocolMessage.Action.presence && message.presence != null) {
+						synchronized (sentPresence) {
+							Collections.addAll(sentPresence, message.presence);
+							sentPresence.notify();
+						}
+					}
+					return true;
+				}
+			});
+
+			ClientOptions opts = createOptions(testVars.keys[0].keyStr);
+			opts.clientId = testClientId1;
+			ably = new AblyRealtime(opts);
+
+			Channel channel = ably.channels.get("protocol_enter_message_format" + testParams.name);
+			/* using testClientId1 */
+			channel.presence.enter(null, null);
+			channel.presence.enterClient(testClientId2);
+
+			synchronized (sentPresence) {
+				while (sentPresence.size() < 2)
+					sentPresence.wait();
+			}
+
+			assertEquals("Verify number of presence messages sent", sentPresence.size(), 2);
+			assertTrue("Verify presence messages follows spec",
+					sentPresence.get(0).action == Action.enter &&
+							sentPresence.get(0).clientId.equals(testClientId1) &&
+							sentPresence.get(1).action == Action.enter &&
+							sentPresence.get(1).clientId.equals(testClientId2)
+			);
+
+			channel.detach();
+			new ChannelWaiter(channel).waitFor(ChannelState.detached);
+
+			try {
+				channel.presence.enterClient("testClient3");
+				fail("Presence.enterClient() shouldn't succeed in detached state");
+			} catch (AblyException e) {}
```

Comment:
I really dislike these blanket exception catches because if another exception is thrown in future (i.e. due to a regression), then this test still passes.  Can you please check the error code and possibly the error message please to make sure it's the right exception being thrown?